p2p/net/swarm/swarm_listen.go

```diff
@@ -6,65 +6,62 @@ import (
 	mconn "github.com/ipfs/go-ipfs/metrics/conn"
 	inet "github.com/ipfs/go-ipfs/p2p/net"
 	conn "github.com/ipfs/go-ipfs/p2p/net/conn"
-	addrutil "github.com/ipfs/go-ipfs/p2p/net/swarm/addr"
+	transport "github.com/ipfs/go-ipfs/p2p/net/transport"
 	lgbl "github.com/ipfs/go-ipfs/util/eventlog/loggables"
 
 	ma "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr"
-	manet "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-multiaddr-net"
 	ps "github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-peerstream"
 	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
-	multierr "github.com/ipfs/go-ipfs/thirdparty/multierr"
 )
 
-// Open listeners for each network the swarm should listen on
-func (s *Swarm) listen(addrs []ma.Multiaddr) error {
+// Open listeners and reuse-dialers for the given addresses
+func (s *Swarm) setupAddresses(addrs []ma.Multiaddr) error {
+	for _, a := range addrs {
+		tpt := s.transportForAddr(a)
+		if tpt == nil {
+			return fmt.Errorf("no transport for address: %s", a)
+		}
 
-	for _, addr := range addrs {
-		if !addrutil.AddrUsable(addr, true) {
-			return fmt.Errorf("cannot use addr: %s", addr)
+		d, err := tpt.Dialer(a, transport.TimeoutOpt(DialTimeout))
+		if err != nil {
+			return err
 		}
-	}
 
-	retErr := multierr.New()
+		s.dialer.AddDialer(d)
 
-	// listen on every address
-	for i, addr := range addrs {
-		err := s.setupListener(addr)
+		list, err := tpt.Listener(a)
 		if err != nil {
-			if retErr.Errors == nil {
-				retErr.Errors = make([]error, len(addrs))
-			}
-			retErr.Errors[i] = err
-			log.Debugf("Failed to listen on: %s - %s", addr, err)
+			return err
```

Comment:
@whyrusleeping again, this _is not_ an error that should stop processing! look at the old code. remember that this _has_ to work when the user's machine moves across having interfaces and transports available. (yes yes we need to reconsider all this and try again when we detect interface changes, but _this_ is going to make ipfs just not work a huge percentage of the time). 

look back through PR for more comments on this.